### PR TITLE
Late binding

### PR DESCRIPTION
**Pull request by asiragusa**
Ext.define('MyApp.controller.Conrtroller', {
    extend: 'Deft.mvc.ViewController',

```
observe: {
    'view.store': [{
        event : "remove",
        fn : "onRecordRemoved",
        lateBinding : true
    }]
},
```

})

### CODE REVIEW

**Comment by asiragusa** [comment]
This commit is more a POC of the concept, the code is not 100% clean everywhere but worth reading ;)

The key concept is that it's impossible with the current version of DeftJS to control events fired before `afterrender` because `control` are binded after such event.

This happens also for Live events, which are supposed to be something similar to ExtJS' `this.control` in Controllers and should be able to listen to every event fired by its controlled components.

The implementation I did completely discards non-live events. This allows the developer to be unaware of the live option that allows to listen to the events fired before `afterrender` (it's awkward for a Framework, it has just to work...)

In the wiki I'm reading the following statement:

> As you can imagine, this is extremely useful in a ViewController that manages a view with dynamically added or removed child items. Keep in mind that this does add some overhead, since DeftJS has to monitor the view continuously. So you don't want to just make all of your control items use live selectors. However, in cases where you need to handle things being dynamically added or removed, it's a very powerful feature.

The phrase 

> DeftJS has to monitor the view continuously

is not really correct, looking at the implementation of Live Events.

The overhead added by LiveEventBus is really lower than the one added by ExtJS' control in EventBus (http://docs.sencha.com/ext-js/4-1/source/EventBus.html#Ext-app-EventBus) as ExtJS monitors EACH event (look at the dispatch method...).

It's true that thousands of controls could slow down the application, as the actual implementation iterates over all the listeners, possibly repeating tests. One idea is to organize listeners by selector to cut lots of comparaisons. This would speed up a lot the algorithm, I'll work it out in the next commit ;)

With this implementation it's also possible to `observe` the view's objects (as a non-singleton injected store). The lateBinding parameter I added in a previous commit becomes unnecessary.

**Comment by brian428** [comment]
Can you rebase so the pull request is up to date with what's in the master branch? Also, I don't see any updated tests to verify the new behavior...did you add any? Finally, have you run it against the existing tests to verify that the existing tests all pass against ExtJS and Touch libraries?

**Comment by asiragusa** [comment]
Ok I'll rebase it. At the beginning it was only a POC. I still dont have run the tests, I'll do it when finished. 

After the last commit I discovered a problem registering the handlers for header and the toolbar elements. The problem is here:

https://github.com/asiragusa/DeftJS/blob/e643fbe05fa4b147a3abf2d81dc1b7d397ebd628/src/coffee/Deft/event/LiveEventListener.coffee#L51

`component.isDescendantOf( @container )` returns false because the component is added to the header or to the toolbar but the bar and the header themselves are still not added to the container.

One idea I have is to inject in the component the `LiveEventListener` class (in https://github.com/asiragusa/DeftJS/blob/e643fbe05fa4b147a3abf2d81dc1b7d397ebd628/src/coffee/Deft/event/LiveEventListener.coffee#L32) and to override the fire method in Component, more or less like ExtJS' EventBus does. What do you think about?

**Comment by asiragusa** [comment]
Now everything works for ExtJS, I have an awkward error in four tests for Sencha Touch, but it's in 

`it('should (when specified within a mixins Object) create an instance of the view controller specified by the target view`controller`property and configure it with a reference to the target view instance when an instance of the target view is created'`

and following, so it's not so important atm.

I have modified some tests, but it's only minor changes. Before each change you will find #TODO: 

Tomorrow I'll rebase the project, but if you want you can begin looking at what I did (there are still possible improvements)

**Comment by asiragusa** [comment]
I merged with latest branch. 

There are 4 tests failing on ExtJS 4.1.1a and below and it's due to the way they test for the presence of event listeners. By the way it's not a problem because this test is obsolete, looking at the way I listen and dispatch the events ;)

Sencha Touch fails also some tests when `Deft.mixin.Controllable` is specified in a mixin array. I don't care a lot as this is becoming deprecated in next versions, but if you want to solve the problem you are free to do it ;)

In a brief summary, this version includes the following:
- The ability to control the events fired before the controlled view's `afterrender`. Events can be listened beginning at the moment when the child view is attached to the current view. It's possible to listen to even the `add` event (the very first fired) for the direct children of the container (those configured by the items config). At this stage it's not possible to listen for the add event in toolbar and header elements (maybe the list is longer...), as they are added to their container before that this is added to the parent's container (ExtJS bug? :) ). However this should not be a problem, as all the other events can be listened
- The live option (in this version) can be considered deprecated. The drawback is that each time a component getter is run, an `Ext.ComponentQuery.query` is executed. It can be worth using this config option to build up a lazy cache of elements whether it's not defined.
- The ability to observe view objects. This is handy when the view's store is not a singleton, thus it can't be injected directly into the controller. IE:

``` javascript
Ext.define('MyController', {
    extend: 'Deft.mvc.ViewController',

    observe: {
        'view.store': [{
            event : "someEvent",
            fn : "onEvent"
        }]
    }
...
});
```
- The ability to inject objects as a constructor config for classes that inherit from `Ext.Component`

``` javascript
Ext.define('MyView', {
    {
        inject : {
            store : 'myStore'
        }
    }
})
```

There is still some work to do to fine tune what I did, however the algorithm I implemented is surely a lot faster than ExtJS' EventBus. For sure it's slower than the actual implementation, but the features it introduces are really important.

I think that people porting their applications from plain ExtJS to Deft JS will appreciate a lot these features, as listening to afterrender is really common and actually impossible to (even with live event listeners).

With this implementation `Deft.mvc.ComponentSelectorListener` can be deleted and some code refactored, as there is now only one way to add event listeners.

Hope that what I have done is useful ;)

**Comment by brian428** [comment]
Just a note that this is probably going to take a while to go through, since it's such a large change set encompassing several separate features (there are roughly 900 lines changed across 11 CoffeeScript files). I didn't want you to think we were ignoring this, it's just going to take some time due to the number of changes.

**Comment by asiragusa** [comment]
Yep, for sure such large changes take a while to apply.

If you have any considerations or doubts don't hesitate to ask, there is a lot of code and improvements can be done. Sharing your ideas can be interesting and speed up things ;)

**Comment by brian428** [comment]
First question: you mentioned that a few tests fail under 4.1.x "due to the way they test for the presence of event listeners." Do you have any idea why they fail now when they used to pass?

**Comment by asiragusa** [comment]
Well, it was hard to catch but there was a bug in LiveEventListener.

Finally the bug is for ExtJS 4.2 and Sencha Touch, `observable.hasListener( eventName )` returns a wrong value.
(https://github.com/deftjs/DeftJS/blob/master/spec/coffee/Deft/mvc/ViewController.coffee#L14)

This made me think of a bug in ExtJS < 4.2 and ST, but it was the opposite...

You will find the bugfix in the next commit

**Comment by brian428** [comment]
Looks like something is pretty wrong when running these changes under Touch. The failures in the Controllable spec actually seem to have nothing to do with specifying the Controllable mixin. If you comment out the mixin, the specs still fail. Looks like specifying the "controller:" property is just being outright ignored, meaning these changes will completely break Touch apps using DeftJS...

**Comment by brian428** [comment]
That said, if I revert your changes to just Deft.mixin.Controllable, it works again. Did your changes to this class actually affect the other features you were adding? Or was this just an attempt to clean up the code? It looks like the latter, but just making sure.

**Comment by asiragusa** [comment]
I did a code clean-up and  an important change. https://github.com/deftjs/DeftJS/pull/73/files#L8R45

The `controller.controlView( @ )` has to be called as soon as possible to be able to listen to all the events, as none has been fired. Calling `controller.controlView` before the `@callParent` doesn't work, as the view is not observable at that moment, so I realized this by hooking the `initComponent` method and it works great for ExtJS. 

That said, I'm not so familiar with ST, so maybe the method to hook is another. Ideas?

BTW tests should be added to verify that the add event / beforerender / afterrender have been caught for the view and the subcomponents, so this tests have to be slightly different from the others, as you can achieve this only by controlling the view you are creating by the `controller` param.

**Comment by brian428** [comment]
Touch has no initComponent() method (yeah, I know). It does have an initialize method, but I'm not sure exactly when it fires or how it relates to Ext's initComponent...

**Comment by asiragusa** [comment]
I'm looking at http://docs.sencha.com/touch/2-0/source/Component.html#Ext-Component

``` javascript

    /**
     * Creates new Component.
     * @param {Object} config The standard configuration object.
     */
    constructor: function(config) {
...
        Ext.ComponentManager.register(me);

        me.initElement();

        me.initConfig(me.initialConfig);

        me.initialize();
```

The component is registered in the `ComponentManager` before the initialize method, so this would not work (as the add event for the view has been already fired)

However, in ST 2, Components inherit directly from `Ext.Evented`, which mixes in `Ext.util.Observable`. The constructor of `Observable` is not called there, nor in `Ext.Component`, so I suppose that you can call the `controller.controlView(@)` directly in the hook of the constructor (but always before the `@callParent`).

That said it is necessary to hook `initComponent` in ExtJS, as it's not possible to attach event handlers in the view before the parent constructor has been called.

However the events we need to hook are beforedestroy and afterrender (https://github.com/asiragusa/DeftJS/blob/lateBinding/src/coffee/Deft/mvc/ViewController.coffee#L212)

this could be achieved also with a LiveEventListener, thus hooking initConfig for ExtJS would be unnecessary.

Thoughts?

**Comment by brian428** [comment]
My instinct is to leave things as they are for the moment. John and I have plenty to go through here without adding another change in. Once we digest what's already here, we can come back around to the option of trying to allow listeners for the earlier view events.

**Comment by asiragusa** [comment]
Well I did it :)

I know that it 'more meat' but it just works ;)

**Comment by asiragusa** [comment]
As a side note, now 

``` javascript

    ###*
    * @private
    ###
    onViewBeforeDestroy: ->
        return @destroy()
```

becomes useless and the event can be binded directly to the destroy method. 

However here there was a little mistake imho, because in ExtJS the destroy process is halted only by returning false during the beforedestroy event. 

Returning undefined ( return; ) in ViewController's destroy  did cancel the view destruction, in contrast with ExtJS philosophy where you have to return false to stop an event firing.

**Comment by asiragusa** [comment]
I moved the init method before calling @createViewObservers, this is useful in case we want to observe objects assigned to the view during the init method.

**Comment by asiragusa** [comment]
I have removed "view observers" and added the `observe` config parameter in `control`. This can be useful in cases like this:

``` javascript
Ext.define('MyController', {
  extend : 'Deft.mvc.ViewController',
  control : {
    view : {
      observe : {
        store : {
           event : 'handler'
        }
      }
    },
    component : {
      selector: 'componentSelector',
      observe : {
        store : {
           event : 'handler'
        }
      }
    }
  }
});

```

I think that the advantages are evident ;)

**Comment by asiragusa** [comment]
Hi guys, news?

**Comment by brian428** [comment]
I set up a branch with these changes merged in (https://github.com/deftjs/DeftJS/tree/__experimental_latebinding). I think the main issue is the sheer size of the changeset. Remember, this pull request:
- Makes some very fundamental changes to the way the framework works
- Contains multiple separate changes or new features
- Touches almost every file
- Alters over a third of the entire codebase

I don't want to discourage you since we do appreciate the effort. You're really the first "external" person to try and contribute in a major way to the framework. But John and I are both pretty busy people, and between work, making sure things keep working with new versions of Touch and ExtJS, the docs, and the mailing list, etc. our time is stretched pretty thin.

I know you've already invested a lot yourself, so I hate to ask, but is there any way you could separate this into some smaller, more targeted pull requests? e.g. one for the changes to the live listeners, one for changes to the injector, one for changes to the Observe feature, etc.?

That's probably an annoying request, but I think it would make our (well, John's really) job a lot easier if the impact of the changes could be assessed in smaller pieces, rather than a single gigantic, intermixed changeset.

Thanks,

Brian

**Comment by asiragusa** [comment]
Hi Brian, 

I understand clearly your problem, but the injector feature takes more or less 10 lines of code and is localized to core/Component. 

The observe feature is based on top of the live listeners. Basically when a view fires the afterrender event I begin observing the objects specified in the config. This also takes very few lines of code.

90% of the changes or more concern the live listeners...

Other ways to help?

**Comment by brian428** [comment]
hmm...if the bulk is in the live event listeners, maybe a text file that goes over what changed in each file and what the rationale was? It wouldn't need to be line-by-line, but having something to look at which explains in some detail what is going on could help. It would be something that might highlight areas to focus on. I'm thinking there are probably certain changes that have more potential concern than others.

**Comment by asiragusa** [comment]
Hi Brian, sorry for being late... I added some comments to the source code.

The most important changes are done in 3 files: LiveEventBus, LiveEventListener and ViewController, other are straightforward changes. I tried to comment better the changes directly in the source code, however i do a small abstract here.

In ViewController:

The main difference is the initializeView that is called as soon as the controlView is called. This allows to setup all the component references and selectors as soon as possible.

I added the component observer feature. This feature is not disruptive and allows us to observe an observable inside a child view in this way:

``` javascript
Ext.define('MyController', {
  extend : 'Deft.mvc.ViewController',
  control : {
    view : {
      observe : {
        store : {
           event : 'handler'
        }
      }
    },
    component : {
      selector: 'componentSelector',
      observe : {
        store : {
           event : 'handler'
        }
      }
    }
  }
});
```

The afterrender / initialize and beforedestroy events are handled like live events.

In LiveEventBus:
- The Listeners data structures becomes an hash, indexed by the selectors. 
- Deft.Component is defined in a separated file
- The register method has been modified to be able to be called by the ViewController, in case the component has been already rendered

In LiveEventListener:

Here is the main change of my patch: the matches method is called for every event fired and not during the register method. This allows to verify if the component is descendant of the main when the event is fired. In this way we can register a live event listener even if the component hasn't been attached to the parent component (thus isDescendantOf would fail).

**Comment by Vaevictus** [comment]
asiragusa just a big thank you as your changes make listening to store events in the controller a piece of cake. Good job!

**Comment by johnyanarella** [comment]
Thanks for the write-up.  We'll be addressing this pull request as part of the upcoming 0.9 release.

**Comment by asiragusa** [comment]
Observable.js should be modified like in my commit.

Basically Ext.dom.Element does not mix in Ext.util.Observable, like in Sencha Touch and the following code won't work otherwise:

``` javascript
Ext.define('MyController', {
    extend: 'Deft.mvc.ViewController',

    control : {
        view: {
            observe: {
                el: {
                    click: 'onClick'
                }
            }
        }
    }
});
```

**Comment by asiragusa** [comment]
Hi Brian, which is the status of the integration with the main branch?

**Comment by brian428** [comment]
I believe John still has this slated to look at as we build up to the 1.0 release.

**Comment by asiragusa** [comment]
I see, can I help in some way to integrate it in the current version?

**Comment by brian428** [comment]
It would probably save John a good bit of time if you could update to the latest master branch code so that the pull request is up to date and more easily merged in.

I know the main change relates to the live event listeners, but I think it would also help if you could basically make this into a few separate commits. Say, one that changes the injector, then another commit that makes the changes to the live listeners, and then another commit that adds the changes to observe. My thinking is that anything that helps John evaluate the different pieces as separately as possible would help. This way, if nothing else John could set a local branch to each revision and step through them.

I'd almost say make it 3 pull requests, but since much of it builds on other changes you're making, that would probably end up being more complicated than helpful, since each pull request would have to be merged into a prior pull request.

John, any thoughts/comments? Would any of these ideas help?

**Comment by brian428** [comment]
And obviously, making sure that once your branch is up to date with master, that all of the tests still pass, etc.

**Comment by asiragusa** [comment]
I'm updating the main branch of my repository: https://github.com/asiragusa/DeftJS, it was easier, and non-disruptive for me to update that branch

**Comment by asiragusa** [comment]
That's it, I tried to separate all the features as much as I could. Please try to integrate them asap in the master branch, as I can't handle to port them each time...

If you have questions on my work, as always, don't hesitate to ask me ;)

**Comment by asiragusa** [comment]
In the last commit that you can find in my master branch I have added the controller inheritance concept.

In my application I have a base view from which many classes inherit from. This base view needs a controller, as their child classes, so I have something like this:

``` javascript
Ext.define('A', {
    controller: 'ControllerA',
    controllerConfig: {
        foo: 'bar'
    }
});

Ext.define('B', {
    extend: 'A',
    controller: 'ControllerB'
})

Ext.define('C', {
    extend: 'A'
    controller: 'ControllerC'
})
```

I need to create one instance of the controllers A and B or A and C for the views B and C, and each controller needs his own controllerConfig, as defined in the view. 

Each controller can be accessed from the view instance with the method getController. If no argument is passed, the top level controller is returned, otherwise the controller with the corresponding class name.

``` javascript
var b = Ext.create('B');

b.getController() // returns the controllerB instance

b.getController('ControllerA') // returns the controllerA instance
```

Getting the controller instance by its class name is maybe not the best choice, in a future version it can be changed with an alias given to the controller.

Another actual caveat is the direct instantiation of the view:

``` javascript
Ext.create('B', {
    controllerConfig: {
        foo: 'bar'
    }
});
```

foo is given as a config parameter to all the controllers. This should be changed in a future version to pass the config parameters to the good controller constructor.

**Comment by brian428** [comment]
Hmm, so are you doing something in the pre-processor that keeps these two controller: properties separate in the extended view? Because right now, I'm fairly sure the child view's controller: would override the parent's controller: declaration. 

I'll have to look at this a bit further, but if I'm following what you're saying, I'm not sure I like this "a view can have multiple parent view controllers" idea, or of changing getController() to have views access separate controller instances. I would think that if view B extends view A, then rather than having two view controllers, it would have one, and ViewControllerB would extend ViewControllerA if desired.

Basically, you may need to explain what you mean here so that everyone is clear. If I'm reading this right and you're talking about having multiple controllers for a view, I'm not sure that's going to fly. That would be a fairly major change to the way things currently work.

**Comment by asiragusa** [comment]
Yep the child view overrides the parent controller property. In the patch I keep track of the controllers, putting them in a list, and then I create an instance for every controller. The diff is not so heavy this time ;)

As you say, there are two possible solutions to keep in a child view the parent's ViewController's functionalities:
- Inheriting from the parent ViewController
- Each view in the hierarchy has his own ViewController

The first solution has an important drawback: we have to know, for each ViewController in the hierarchy of A, that we have to extend A's ViewController, otherwise we are loosing the A ViewController's functionalities.

The second solution is more transparent. Each view in the hierarchy has his own controller and, if we extend a view, we don't have to worry about A's ViewController, it's just working because A carries out his ViewController. 

The second solution, as you say, creates multiple instances of ViewControllers for the same view, each one controlling a different aspect of it. The slight drawback of this solution is that view.getController needs to know the type of the controller we want. However this is not a major problem, as it's not a good practice to call directly a ViewController's method from the outside, and we often know to which kind of ViewController we are referring.

**Comment by brian428** [comment]
OK, so just a couple of things:

First, I love your enthusiasm and I'm really not trying to blunt your good intentions, but my worry is that this is getting a bit out of control. There's already a lot (a LOT) that needs to be gone over in what you've already got. Every time you make another change, it's really just making it that much harder to go over and assess. So my first bit of advise would be to create new issues for these ideas so that they can be discussed before you take spend time actually coding it. If you really want to code it up, that's fine, but I'd say do it in a separate branch so we can keep some sort of compartmentalization around these proposed changes. 

On a more fundamental level, Sencha has retroactively made class pre-processors private. They already have optimizer logic in the new builds of Cmd which completely breaks class pre-processors. Which means we're very likely going to have to return to using mixins for a lot of Deft's functionality. Which unfortunately means that approaches like this (which rely on class creation hooks) are probably not going to be an option going forward. 

So, I'm all for having these kinds of discussions about new features, but we need to figure out what is going to happen before someone runs off and just does it. Again, I'm not trying to squash your ideas, but we've got to maintain some order when managing changes, particularly sweeping changes. Make sense?

**Comment by asiragusa** [comment]
Yep, it makes sense ;) But I can't be stuck with my application, so I have to write down the functionality. However for most of the functionalities I've added, it didn't take too long to write them down.

Separating them in different branches is hard, or it doubles or triples the work I need to do, as I need the whole package to be working. 

I understand that it takes longtime to integrate my code, but my last major commit was 5 months ago, and merging the  major changes you have done recently took me a lot of time.

If you look at the commits I have done in the new branch, I've tried there to separate all the changes as much as I can, and you will find that each time just one file or two are involved for each feature.

That said, the approach I've taken for the last commit follows what have been done for the Injectors, so the solution that has to be taken if it's decided to come back to mixins will be equivalent for both classes.

Speeding up the integration of my changes would solve all the matter as well ;)

**Comment by brian428** [comment]
Yep, there were just other major updates that had to take priority, like the Promises API and converting the entire project over to the Sencha package manager. Also, as I said, the fact that Sencha is killing the whole idea of class pre-processors is probably going to have a major impact on both the current code and what you've done in your branch. We're still not entirely sure what the full scope of that decision is going to mean.

**Comment by asiragusa** [comment]
Yep Brian, but as I said before, I almost copied the solution from the one used in the injectors, then if it's decided to migrate to mixin, the porting will be equivalent, and straightforward, as it's only few lines of code that change.

The new version of the LiveEventListeners is not impacted at all by this decision, as it doesn't rely on class preprocessors. Have a look at Deft.mixin.Controllable, in the past commits I just anticipated the call to controller.controlView( @ )

Try to use my fork for one of you projects, you will notice that it really simplifies some tasks, as you don't need anymore to work around to some limitations of the Framework.

As a side note, I really don't like the idea ok killing the class preprocessors. Recently I did a project where I needed to share the maximum of the source code between ExtJS and Sencha Touch, and they helped me a lot to sort the mess that Sencha has done with both Frameworks. There is a real gain in term of performance to kill class preprocessors?

**Comment by brian428** [comment]
The real problem is that if we have to ditch the pre-processors, this sort of logic that allows for dealing with inherited properties (either to keep track of them or to merge them) is going to be difficult (impossible?) without the ability to hook into the process before the prototypes of the superclasses are merged. Which means migrating this to mixins may result in actually losing that functionality (for injections, observe, etc.). I'm not sure how that can be done with mixins alone.

I also am pretty iffy on the decision to hide the pre-processors. Further, it just bad form to _retroactively_ make them private. You don't take something that was public and just suddenly kill it. But I guess not many people outside Sencha actually use them, so they thought (incorrectly) that this would be ok to do.

My guess is that the Cmd optimizer has some awareness of the built-in class processors, but that letting external code hook into that process would be more difficult to account for. I'm not really sure what the real-world performance difference is, but it almost doesn't matter. Right now people using Deft have to manually go disable the optimizer, which is obviously not viable going forward.

**Comment by asiragusa** [comment]
I totally agree with you, class pre-processors are fundamental for DeftJS and hiding them is an awful idea. Have you tried contacting Sencha developers about that?

I will try to dig into the source code in the next days to see if I can find out some workaround ;)

**Comment by asiragusa** [comment]
I managed to fix the bug introduced with the ExtJS optimizations, in the last commit there is the bugfix. 

It was quite tricky to solve, I hope that the solution I found would work across different versions, even if I'm touching at private classes.

Don't hesitate to ask me questions about the fix, I'll explain better if needed.
